### PR TITLE
Feature/pdct 1628 handle slugs for documents with the same title

### PR DESCRIPTION
**Pull request by annaCPR**
# Description
- keep track of document slugs created as part of bulk import so that new slugs created can be validated against them
- increase the length of a slug suffix from 4 to 6 to minimise the chance of getting a collision
- refactor ingest service to surface errors on commit to DB and improve logging

Please include:

- a summary of the changes
- links to any related issue/ticket
- any additional relevant motivation and context
- details of any dependency updates that are required for this change

## Proposed version

Please select the option below that is most relevant from the list below. This
will be used to generate the next tag version name during auto-tagging.

- [ ] Skip auto-tagging
- [x] Patch
- [ ] Minor version
- [ ] Major version

Visit the [Semver website](https://semver.org/#summary) to understand the
difference between `MAJOR`, `MINOR`, and `PATCH` versions.

Notes:

- If none of these options are selected, auto-tagging will fail
- Where multiple options are selected, the most senior option ticked will be
  used -- e.g. Major > Minor > Patch
- If you are selecting the version in the list above using the textbox, make
  sure your selected option is marked `[x]` with no spaces in between the
  brackets and the `x`

## Type of change

Please select the option(s) below that are most relevant:

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] GitHub workflow update
- [ ] Documentation update
- [ ] Remove legacy code
- [ ] Dependency update

## How Has This Been Tested?

Please describe the tests that you added to verify your changes.

## Reviewer Checklist

- [ ] **DB_CLIENT DEPENDENCY IS ON THE LATEST VERSION**
- [ ] The PR represents a single feature (small drive-by fixes are also ok)
- [ ] The PR includes tests that are sufficient for the level of risk
- [ ] The code is sufficiently commented, particularly in hard-to-understand areas
- [ ] Any required documentation updates have been made
- [ ] Any TODOs added are captured in future tickets
- [ ] No FIXMEs remain

### CODE REVIEW

File: app/repository/helpers.py [hunk 1]
```diff
@@ -10,40 +10,62 @@
...
+    :return str: The slug generated.
     """
-    lookup = set([cast(str, n) for n in db.query(Slug.name).all()])
     base = slugify(str(title))
 
     # TODO: try to extend suffix length if attempts are exhausted
     suffix = str(uuid4())[:suffix_length]
     count = 0
-    while (slug := f"{base}_{suffix}") in lookup:
+    while (slug := f"{base}_{suffix}") in existing_slugs:
```
Comment (diversemix) [inline app/repository/helpers.py:32]: don't we want to check  a combination of `lookup` (slugs from the db) and `existing` (slugs part of the transaction)

File: app/repository/helpers.py [hunk 2]
```diff
@@ -10,40 +10,62 @@
...
+    :param Session db: The connection to the db
+    :param str title: The title or name of the object you wish to slugify
+    :param int attempts: The number of attempt to generate a unique slug before
+    failing, defaults to 100
+    :param int suffix_length: The suffix to produce uniqueness, defaults to 4
+    :param Optional[set[str]] created_slugs: A set of slugs created in the context within which
+    this function runs that have not yet been committed to the DB
+    :return str: the slug generated
+    """
+    saved_slugs = set([cast(str, n) for n in db.query(Slug.name).all()])
```
Comment (diversemix) [inline app/repository/helpers.py:62]: Probably worth a note here - to improve this code / reference to a ticket

**Review by diversemix — APPROVED** [review]
lgtm

File: app/repository/helpers.py (hunk 1 above)
Comment (annaCPR) [reply app/repository/helpers.py:32]: Yes, that's why they get combined in the function that calls this one so this function still only gets a single set with all slugs. See generate_slug()

File: app/repository/helpers.py (hunk 2 above)
Comment (annaCPR) [reply app/repository/helpers.py:62]: Can you clarify what you mean here please :) ?

**Pull request merged**